### PR TITLE
Improve AST safety parsing error message

### LINKED ISSUE #2178: Provide a more useful error when parsing fails during AST safety checks
### Summary & why

Currently when Black encounters a SyntaxError while parsing during its AST safety check, it tries older versions of Python to see if it will parse then under its grammar (i.e. the syntax - which is an oversimplification but generally words). In other words, if you have a file that is valid Python 2.7 code and are running on Python 3.8, Black will first try will try Python 3.8 which fails, then 3.7 which also fails, then 3.6, 3.5, and finally 2.7. Since 2.7 is last version it tried parsing with, the error message from that version is the one ultimately emitted to the user.

The behaviour described above logically makes sense but is of limited use and sometimes even annoyance of the end-user since chances are that their file is supposed to be valid Python 3, so a message for Python 2 ain't useful (and can be downright confusing). 

### Solutions & technical details 

We should change Black so when a SyntaxError is encountered while parsing during the AST safety checks it reparses with the latest Python it can try. Then we can take its error message and emit that one to the user, and voilà, a much more useful error message.  

Another option (from @ichard26 personally) is that we could record the error message of the latest Python parse attempt when it fails and then if all of the parse attempts fail, emit the recorded error message instead of freshest error message (probably Python 2). 

One slight issue is that we emit an error telling users if they need Python 2 support they should reinstall Black with the `python2` extra when we try Python 2 parsing (which fails due to missing dependencies). This could be downgraded to a warning so the rest of this enhancement can still do its thing. Or it could just be eliminated which is probably fine since Python 2 is basically dead. 

To potential contributors: 

- [Checkout our contributing docs to get started](https://github.com/psf/black/blob/master/CONTRIBUTING.md). If you need help, feel free to reach out!
- You'll need to definitely touch this part of the codebase: https://github.com/psf/black/blob/14c76e89716b5b53c97ece80bb935ea956b7dd89/src/black/__init__.py#L6422-L6443
- You'll probably also need to add / modify some tests in `tests/` 
- And finally add a changelog entry

\- Richard (@ichard26) (added for anyone looking to contribute)

---

## Original Report

This one had me scratching my head for a while.

**How to reproduce**

Feed the following into black with -t py37:

```
reported_error = f""

e=""
# Actual error
malformed_fstring = f"{e}  ({type(e).__name__], {e.args})"

```
Please note the typo in `malformed_fstring`: I put a square closing bracket "]" in place of a curly closing bracket "}".
Please also note that if the e variable is not set, the error does not show up.

**Expected result**

I expected that black would form a usable error message about "malformed_fstring having unmatched parenteses" or some such thing.

**Actual result**

```
black -l 999 -t py37 "test.py"
error: cannot format /tmp/test.py: cannot use --safe with this file; failed to parse source file.  AST error message: invalid syntax (<unknown>, line 1)
Oh no! 💥 💔 💥
1 files left unchanged, 1 file failed to reformat.
```

Please note that it reports an error on line 1, even if the error is on line 5.

**Environment**

```
$ black --version
black, version 20.8b1
$ python --version
Python 3.7.4
$ uname -a
Linux y8 4.19.0-16-amd64 #1 SMP Debian 4.19.181-1 (2021-03-19) x86_64 GNU/Linux

```
**Possible causes**

After digging for a bit it it seems this happens when black fails to use the AST for python 3.x for some reason and falls back to trying the AST for 2.x which does not support f-strings at all and then barfs on the very first instance of an f-string. In the future, I will take note that whenever black barfs on an f-string, it is because it is parsing thje file as 2.x because it completely failed to parse as 3.x, and instead of looking at the f-string mentioned, start going all out Newton on the rest of the file.

As for why this exact string produces this outcome I can only speculate it has something to do with the "type" keyword being misinterpreted somehow.

### PR DESCRIPTION
**Pull request by felix-hilden**
Closes #2178, alternative to #2218: This PR continues on the work of Hasan Ramezani in the alternative PR to improve the error messages when parsing formatted AST. His original commit is till at the beginning of the branch, which I then modified (Thank you for the initial push!)

A number of non-obvious details and questions:
- As suggested by Richard in the linked issue, I removed the error messages (and a test) for Python 2 parsing
- Previously versions 3.6 and 3.7 were parsed twice, first by the builtin `ast` and then by `typed_ast`. On [Discord](https://discord.com/channels/267624335836053506/846434317021741086/850083982459535400) we determined this was an accident so I removed it.
- The parsing was refactored a bit for clearer flow. It's now two functions: one for parsing a single version, and the wrapping function for constructing a list of versions to parse and then doing it. Probably easier to view the source than the diff on that one.
- Mypy linting fails, even though none of the type signatures were changed. Should we ignore it or is there some trickery that could be done to make Mypy recognise the types?

Everything's in separate commits for an easier review. I'll leave this as a draft at least until the Mypy issue has been resolved.

### CODE REVIEW

**Comment by felix-hilden** [comment]
Oops. Didn't realise tests wouldn't run in a draft!

**Comment by JelleZijlstra** [comment]
Not sure what's going on with the mypy error. Can look in more detail in a few hours.

**Comment by felix-hilden** [comment]
Also, target versions include Pythons all the way from 3.3. Should they be here too?

File: src/black/parsing.py
Comment (felix-hilden) [inline src/black/parsing.py:115]: Also, target versions include Pythons all the way from 3.3. Should they be here (and in the version list below) too?

**Comment by ichard26** [comment]
I said this on PyDis but I'll say it here for the record and also just in case someone misses it:

> @[felix] I just looked into it and the mypy error seems to be about how ast.parse only started to accept feature_version in 3.8 and higher

> ast.parse(source, filename='<unknown>', mode='exec', *, type_comments=False, feature_version=None)
> [...]
> Changed in version 3.8: Added type_comments, mode='func_type' and feature_version.

> And mypy isn't smart enough to know that ast will only be used on versions where feature_version is supported :slight_smile:
 
The easiest solution would be just adding an assert that makes sure we're running on 3.8 in that branch. Other than that, I can't think of a solution that wouldn't require redoing and restructuring the patch which is kinda defeating the point :/

**Comment by felix-hilden** [comment]
Thanks for digging around! I see now why the error happens only now. So a couple of things we could do:

- `# type: ignore`, since we know that the calling code has made sure we are running on 3.8+ (tested and works)
- Assert or redundantly test in the if `sys.version_info` once more (tested and works)
- Revert the refactoring, if you're not too keen on having it (I know it's only a slight improvement)

I'll leave it up to you!

File: src/black/parsing.py
Comment (ambv) [reply src/black/parsing.py:115]: Looks like the previous behavior was mistakenly parsing 3.3-3.5 with ast27. And if it was somehow deliberate the logic fails to explain it so let's remove that special case and try to parse 3.3-3.5 like you're suggesting. Good find!

Comment (felix-hilden) [reply src/black/parsing.py:115]: I had to go and verify myself how the versions were parsed again. Here's a summary and new versions we'd like to have:
```
typed | sysver | parsed                              | new parsed
-----------------------------------------------------------------------
 yes  | >= 3.8 | ast: sysver down to 3.5, typed: 2.7 | ast: sysver down to 3.3, typed: 2.7
 yes  |  < 3.8 | typed: 3.7, 3.6, 2.7                | typed: 3.7 down to 3.3, 2.7
  no  | >= 3.8 | ast: sysver down to 3.5             | ast: sysver down to 3.3
  no  |  < 3.8 | error on import
```
So what I've understood is that the builtin ast is good for all versions except 2.7 if we're above 3.8, but below it we need `typed_ast`. I think I understood it differently when I first refactored, so I'll change it if I get a nod of approval that this time around it's solid 😄 Thanks for the patience!

Comment (felix-hilden) [reply src/black/parsing.py:115]: Also: would we like to tie the parsed version to target versions? I.e. when we add or remove something from the enumeration, we won't have to change this part of the code.

Comment (ichard26) [reply src/black/parsing.py:115]: IIRC we need typed_ast for <3.8 because the built-in `ast` module has limited feature_version support. In other words, using `ast` <3.8 would force Black to be able to only parse using the syntax of the runtime version. I'm not entirely sure though, I would strongly encourage either putting this information in a comment or the dev documentation because it's been like a year of me being here and I still don't 100% understand this area.

Comment (ichard26) [reply src/black/parsing.py:115]: Also, apparently `ast` and `typed_ast` have feature_version minimums of (3, 4): 

> The lowest supported version is (3, 4); the highest is sys.version_info[0:2].

https://docs.python.org/3/library/ast.html#ast.parse

> Currently only fully supported for Python 3.5+ with partial support for Python 3.4. So, feature_version=3 or less are all equivalent to feature_version=4.

https://github.com/python/typed_ast/blob/8eed936014f81a55a3e17310629c40c0203327c3/typed_ast/ast3.py#L50-L54

The good news is that it really doesn't matter since 3.3 and 3.4 seem to be practically the same syntax / grammar wise:
https://github.com/psf/black/blob/aa31a117b16ed849710cf13cea5c064839beb29e/src/black/mode.py#L47-L48

Comment (felix-hilden) [reply src/black/parsing.py:115]: That makes sense, thank you! I'll shoot changes later today and hopefully we can get this forward!

Comment (felix-hilden) [reply src/black/parsing.py]: Seems like the builting `ast` also accepts feature versions below 3.4 as long as they are above 3. Also, I didn't link the code to `TargetVersion`, because assuming the correct enumeration values for minors or parsing the names felt hacky 😅

**Comment by felix-hilden** [comment]
I corrected the parsing logic now! I think this could be ready for another round of review (hopefully the last as well) 👌

File: src/black/parsing.py
Comment (ichard26) [inline src/black/parsing.py:124]: Technically a bigger concern is the fact 3.10 syntax is not supported but oh well ;)

**Review by ichard26 — APPROVED** [review]
I don't see any issue with this and it's a nice readability improvement while implementing a nice enhancement. Thank you!

File: src/black/parsing.py
Comment (felix-hilden) [reply src/black/parsing.py:113]: Yeah, I think I like this a bit more so let's stick with it unless someone really wants more detail. Once again we should definitely continue the discussion on #2165 or even make a dedicated issue diving into exactly why `typed_ast` is needed. But I think this will do for the current PR.

File: src/black/parsing.py
Comment (felix-hilden) [reply src/black/parsing.py:124]: Yeah 😅

**Review by JelleZijlstra — APPROVED** [review]

**Pull request merged**
